cloudimg/aws.py

```diff
@@ -973,6 +994,19 @@ def delete(self, metadata):
             else:
                 deleted_snapshot_id = self.delete_snapshot(snapshot)
 
+        # If container is specified lets try to find if there's an object
+        # tied to the image and delete it
+        if metadata.container:
```

Comment:
I'm concerned about the relationship between images and objects.

"When I delete an image, I want to delete the associated object" seems to be valid if and only if there is a 1:1 relationship, i.e. there is never a case where an object can be referenced by more than one image.  If an object is referenced by more than one image, then this change would mean deleting one image could break some other images, right?

Are you sure there's always a 1:1 relationship?

I'm looking at `AWSService.publish` right now and it certainly seems like that code is at least *capable* of creating situations where multiple images share an object, as it definitely accounts for the possibility that the object already exists and avoids uploading it in such a case.

The change seems to create a mismatch where:

- the "publish" side of the code can create multiple images referencing the same object
- the "delete" side of the code assumes there's a 1:1 image-object relationship, so that when deleting an image, it's OK to delete the image's object

That is a very dangerous API.  I know that `container` in delete metadata is optional, so you can say "we won't set that except in cases where we know it's safe", but even if you have that in mind I think this is setting up a trap for future developers - and there's not even a warning in the docs about this.

If it's important to fix the problem of useless objects remaining in S3, I would rather see a smarter way to deal with it which somehow ensures that deleting one image will not delete objects referenced by other images.

Reply:
Do you know if it matters if the the file object is available? Since the image registered uses the snapshot. Seems like the code only associates the image with a snapshot and I can't find any information on tying a snapshot with an image object.

Reply:
Having reviewed the code and the APIs, I came to the conclusion that no, it probably doesn't matter.

It seems like the object in S3 is used during the EBS "snapshot import" task, but this doesn't create a persistent reference to the object.  One object can be used for multiple image imports though (e.g. one "access" and one "hourly" image).  I don't know if the object is used for anything after the snapshot import.

So, this probably isn't as dangerous as I'd thought.

Isn't the lifecycle of these objects kind of messed up though?  If we are saying that the S3 object is used just to import the snapshot and is not otherwise needed: then why would it make sense to allow it to continue existing until the image and snapshot are later deleted?  Aren't some images never deleted, meaning we'd never clean up the objects for those?

If the purpose of the uploaded objects is only to be imported as EBS snapshots, which is expected to happen soon after the upload, then I wonder if this would be better handled by having S3 automatically expire and delete the objects (e.g. an S3 lifecycle configuration to expire objects older than 30 days).  That would also clean up objects for images which _aren't_ deleted, leftover objects from interrupted and never retried pushes, etc.

Reply:
Agreed on the lifecycle of objects (especially the image registration process) being odd. I was also wondering over the weekend if it would be better suited to add an s3 lifecycle config. Do you think it makes sense to drop this and shift to setting the lifecycle of the s3 bucket?

Reply:
Based on what I know now, pursuing the S3 lifecycle configuration sounds like a good idea.  Possibly also adding some comments in `publish`, around where the S3 object is uploaded, explaining what's expected to happen to the object afterward.

What's still a bit mysterious to me is whether the specific bucket(s) we're using have also been used for other things - meaning our lifecycle config would have to take care to avoid expiring those other things.